manage.py

```diff
@@ -39,55 +35,128 @@ def create_or_mask(N,ids):
+    #   that rank controllable nodes
+    # budgets - iterable of integers defining max budget
+    #
+    # return
+    # 
+    # iterable of integer-vectors that go up to the (max) budget
+    return (ctl[:b] for ctl, b in zip(ranks, budgets))
+
+def manage_nodes(A, rank_opts,manage_opts):
+    # management of a network by k managers, potentially cooperative
+    #
+    # A is the adjacency matrix
+    # rank_opts['random'] - select node uniformly at random
+    # rank_opts['evec'] - select node by eigenvector centrality
+    # rank_opts['distributed_evec'] - select node by distributed eigenvector centralities
+    # rank_opts['custom'] - allow input of any function with any arguments
+    # rank_opts['info_id_sets'] -  iterable of length k, each element defines
+    #                the information that manager k knows 
+    # rank_opts['control_id_sets'] -  iterable of length k, each element defines
+    #                the nodes that manager k controls 
+    # rank_opts['budget'] -  iterable of length k, with each element an integer defining maximum budget
+    # rank_opts['manage_id_sets'] -  iterable of length k, with each element an integer defining maximum budget
+    #
+    # manage_opts['num_managers'] - num managers (optional) if not present set to len control_id_sets
+    # manage_opts['online'] - if true, adaptively compute management during selection, otherwise
+    #                         compute management from inital state of network
+    #
+    # return AA final adjacency matrix after management, 
+    #        lams vector of dominant eigenvalues during management
+    # 
+
+    budgets = rank_opts['budget']
+    B = int(np.sum(budgets))
+    try:
+        K = manage_opts['num_managers']
+    except KeyError:
+        K = len(budgets)
+    Ks = np.arange(0, B, K, dtype='int')
+    info_to_rank = get_info_to_rank(rank_opts)
+    info_to_control = get_info_to_control(info_to_rank, rank_opts)
+
     if manage_opts['online']==False:
-        ranks = get_node_ranks(A,rank_opts)
-        # OR swap so this retuns a manager-column matrix with ranks in descending order
-        # so add for loop here that encodes info logic of rankings (pooled or not)
-
-        # later down... separate for loop with budget etc and the function that
-        # decreases weight would go through each manager iteratively and they choose
-        # which to manage -- intersect teh manager patches w/ full ranks (subset managed)
-        #print(ranks[:4])
-        lams = np.zeros(len(Ks))
+        # compute ranks just once
+        # logic here is to get a manager-length iterable that ranks in descending order
+        # the nodes that will be managed up to the max budget
+        # we then 'flatten' those in a turn based way (all managers 1st choices occur
+        # before any managers 2nd chioces)
+        ranks = info_to_control(A)
+        management_plans = manage_for_budget(ranks, budgets)
+
+        global_rank = flatten_by_turns(management_plans)
```

Comment:
global rank an iterable over full set of nodes that will be managed at max total budget